### PR TITLE
feat: cache .next/cache between builds

### PR DESCRIPTION
**Pull request by lindsaylevine**
Fixes https://github.com/netlify/netlify-plugin-nextjs/issues/63

questions:
1. how will this behave with the pizzafox plugin? answer: works
2. how can this be tested within the plugin, if at all? answer: mock/spy

### CODE REVIEW

File: helpers/cacheBuild.js
Comment (ehmicky) [inline helpers/cacheBuild.js:5]: Instead of using a `buildDir` input, couldn't users use our `base` configuration property instead?
This changes the current directory of the plugin's process, which is used by default to resolve the file paths passed to `utils.cache.save()` and `utils.cache.restore()`.

File: helpers/cacheBuild.js
Comment (ehmicky) [inline helpers/cacheBuild.js:7]: Out of curiosity, when users Next.js directory is not `{buildDir}/.next`, is it because:
  - The directory is not named `.next`?
  - Or the directory is named `.next` but is not located in `buildDir`?

File: helpers/cacheBuild.js
Comment (lindsaylevine) [reply helpers/cacheBuild.js:5]: hm i suppose so, though i know *as is* (with current monorepo support) the base configuration doesn't work for all monorepo use cases - which i guess begs the qestion, is that why pizzafox's plugin accepts buildDir as input? 🤔

File: helpers/cacheBuild.js
Comment (lindsaylevine) [reply helpers/cacheBuild.js:7]: the first: directory is not named .next

https://nextjs.org/docs/api-reference/next.config.js/setting-a-custom-build-directory

thats why i wrote in my the PR desc that i technically should get distDir from nextConfig, but i just want to wait for #182 to be merged first (because these 2 PRs conflict)

File: helpers/cacheBuild.js
Comment (ehmicky) [reply helpers/cacheBuild.js:5]: The `base` property is definitely more of a patch than a real solution for many monorepo setups.

In this case though, I think the `buildDir` input would use the same strategy as the `base` property. In essence, it would behave the same by changing the base directory to find `.next`.

However, please feel free to discard this comment if you think this might be a useful input. :+1: 

If we do though, I would personally suggest renaming it to `base` because `buildDir` is a somewhat internal name, and to reflect the similar configuration property with the same name.

File: helpers/cacheBuild.js
Comment (ehmicky) [reply helpers/cacheBuild.js:7]: Thanks for the explanation @lindsaylevine 

> i technically should get distDir from nextConfig, but i just want to wait for #182 to be merged first (because these 2 PRs conflict)

After we do this, would there be any use cases left for the `distDir` input?

File: index.js
Comment (ehmicky) [inline index.js:64]: We should add the new `inputs` to `manifest.yml`.

Comment (lindsaylevine) [reply index.js:64]: RIGHT DUH

File: helpers/cacheBuild.js
Comment (lindsaylevine) [reply helpers/cacheBuild.js:7]: nope, none :)

File: helpers/cacheBuild.js
Comment (lindsaylevine) [reply helpers/cacheBuild.js:5]: ok no i can get rid of buildDir (or `base`) so we won't have any inputs after all! if people end up needing it, we can always add it back in :)

**Comment by lindsaylevine** [comment]
@ehmicky any specific thoughts on these questions in the PR desc? 🙏 

how will this behave with the pizzafox plugin?
how can this be tested within the plugin, if at all?

**Comment by ehmicky** [comment]
The cache utility deletes the destination file/directory before both saving and restoring. So I believe it should be idempotent, and running both `netlify-plugin-cache-nextjs` and the logic in this PR should work.

However, I would suggest manually testing it before merging this PR. The production behavior of the caching utility is different from local builds, so I would suggest testing both (or at least production). If it turned out to be incompatible, I would suggest:
 - Detecting whether `netlify-plugin-cache-nextjs` is used. This could be done by using `require('netlify-plugin-cache-nextjs')`.
 - If detected, the logic of this PR could be a noop.

When it comes to testing it, I would suggest using a fixture directory then:
  - Delete `.next/` if already exists (from running the test previously)
  - Create `.next/` with a `.next/dummy` in it
  - Run the plugin
  - Delete `.next/` and `.next/dummy`
  - Run the plugin again
  - Assert that `.next/dummy` exists
  - Delete `.next/` and `.next/dummy`

**Comment by lindsaylevine** [comment]
i manually tested. as for automated tests, i'm not sure how we'd "run the plugin" with the real valid utils.cache hook?

**Comment by ehmicky** [comment]
The above was assuming the tests would be run using `@netlify/build` programmatically. This is more robust, but also more complicated to implement.

Another option would be to spy on the mocked `utils.cache.*` methods, for example using `jest.spyOn()`. This would be simpler to implement, but would also not catch every type of bug.

**Review by ehmicky — APPROVED** [review]

**Pull request merged**
